packages/ai-ide/src/browser/workspace-functions.ts

```diff
@@ -324,3 +331,111 @@ export class GetWorkspaceFileList implements ToolProvider {
         return result;
     }
 }
+
+@injectable()
+export class FileDiagonsticProvider implements ToolProvider {
+    static ID = GET_FILE_DIAGNOSTICS_ID;
+
+    @inject(WorkspaceFunctionScope)
+    protected readonly workspaceScope: WorkspaceFunctionScope;
+
+    @inject(ProblemManager)
+    protected readonly problemManager: ProblemManager;
+
+    @inject(MonacoTextModelService)
+    protected readonly modelService: MonacoTextModelService;
+
+    @inject(OpenerService)
+    protected readonly openerService: OpenerService;
+
+    getTool(): ToolRequest {
+        return {
+            id: FileDiagonsticProvider.ID,
+            name: FileDiagonsticProvider.ID,
+            description:
+                'A function to retrieve diagnostics associated with a specific file in the workspace. It will return a list of problems that includes the surrounding text \
+            a message describing the problem, and optionally a code and a codeDescription field describing that code.',
+            parameters: {
+                type: 'object',
+                properties: {
+                    file: {
+                        type: 'string',
+                        description: `The relative path to the target file within the workspace. This path is resolved from the workspace root, and only files within the workspace 
+                        boundaries are accessible. Attempting to access paths outside the workspace will result in an error.`,
+                    }
+                },
+                required: ['file']
+            },
+            handler: async arg => {
+                const { file } = JSON.parse(arg);
+                let workspaceRoot;
+                try {
+                    workspaceRoot = await this.workspaceScope.getWorkspaceRoot();
+                } catch (error) {
+                    return JSON.stringify({ error: error.message });
+                }
+
+                const targetUri = workspaceRoot.resolve(file);
+                this.workspaceScope.ensureWithinWorkspace(targetUri, workspaceRoot);
+                return this.getDiagnosticsForFile(targetUri);
+            }
+        };
+    }
+
+    protected async getDiagnosticsForFile(uri: URI): Promise<string> {
+        const toDispose: Disposable[] = [];
+        try {
+            let markers = this.problemManager.findMarkers({ uri });
+            if (markers.length === 0) {
+                // Open editor to ensure that the language services are active.
+                await open(this.openerService, uri);
+                // Give some time to fetch problems in a newly opened editor.
+                await new Promise<void>(res => {
+                    setTimeout(() => res, 5000);
```

Comment:
Is this number based on experience? How sure are we, that we wait long enough?

Reply:
Reasonably confident. Cline, for example, waits 3.5 seconds for new diagnostics. The main edge case is if opening the file triggers the initialization of a given language server / service's index because nothing else has been opened yet. That's likely to be fairly unusual, but we could return a descriptive error rather than just an empty array so that the agent sees (and relays) a message that retrying may work.